File: packages/zsh-core-tooldef/src/tools/docs.ts
// MIRRORED-IN: zshref-rs/src/tools/docs.rs

import type { DocCorpus } from "@carlwr/zsh-core"
import { renderDoc } from "@carlwr/zsh-core/render"
import {
  lookupRaw,
  type ResolverFeedback,
  resolverFeedback,
} from "@carlwr/zsh-core/resolver"
import {
  classifyOrder,
  type DocCategory,
  type DocPieceId,
  type DocRecordMap,
  docSubKind,
} from "@carlwr/zsh-core/taxonomy"
import type { Documented } from "@carlwr/zsh-core/types"
import { buildToolDef, type SchemaShape, type ToolDef } from "../tool-defs.ts"
import { docsProse } from "./prose.ts"
import { display } from "./shared/doc-display.ts"
import type { BaseMatch } from "./shared/entries.ts"
import {
  categoryShape,
  type Envelope,
  isValidCategory,
  mkEnvelope,
} from "./shared/envelope.ts"
import { mkOutputSchema } from "./shared/output-schema.ts"

export interface DocsInput {
  readonly key: string
  readonly category?: DocCategory
}

export interface DocsMatch extends BaseMatch {
  readonly mdBody: string
  /** Optional per-category resolver feedback (e.g. `{ kind: "input-negated" }` when an option was reached via `NO_`-stripping). Absent when there is nothing to surface. */
  readonly feedback?: ResolverFeedback
}

export type DocsResult = Envelope<DocsMatch>

function formatMatch(
  corpus: DocCorpus,
  pid: DocPieceId,
  key: string,
): DocsMatch {
  const rec = corpus[pid.category].get(pid.id as never) as
    | DocRecordMap[typeof pid.category]
    | undefined
  if (!rec) {
    throw new Error(
      `docs: corpus lookup miss for ${pid.category}:${pid.id} — resolver returned an id not present in the corpus map.`,
    )
  }
  const getSubKind = docSubKind[pid.category] as (
    d: DocRecordMap[typeof pid.category],
  ) => string | undefined
  const subKind = getSubKind(rec)
  const fb = resolverFeedback(corpus, pid.category, key)
  return {
    category: pid.category,
    id: pid.id as string,
    display: display(pid.category, rec),
    mdBody: renderDoc(corpus, pid),
    ...(subKind !== undefined ? { subKind } : {}),
    ...(fb !== undefined ? { feedback: fb } : {}),
  }
}

/**
 * Look up docs for a zsh key. With `category`, one category (0–1
 * matches). Without, walks `classifyOrder` and returns one match per
 * resolving category — usually 0–1, sometimes 2 when overlap categories
 * both resolve (`for`, `nocorrect`, etc.).
 *
 * Resolution is `lookupRaw` (direct ∥ resolver, direct preferred). Pure; no IO.
 */
export function docs(corpus: DocCorpus, input: DocsInput): DocsResult {
  const key = input.key
  if (key.trim().length === 0) return mkEnvelope<DocsMatch>([])

  if (input.category !== undefined) {
    if (!isValidCategory(input.category)) return mkEnvelope<DocsMatch>([])
    const pid = lookupRaw(corpus, input.category, key)
    if (!pid) return mkEnvelope<DocsMatch>([])
    return mkEnvelope([formatMatch(corpus, pid, key)])
  }

  const matches: DocsMatch[] = []
  for (const cat of classifyOrder) {
    const pid = lookupRaw(corpus, cat, key)
    if (pid?.category === "history_expn" && !isHistoryEvent(corpus, pid))
      continue
    if (pid) matches.push(formatMatch(corpus, pid, key))
  }
  return mkEnvelope(matches)
}

function isHistoryEvent(corpus: DocCorpus, pid: DocPieceId): boolean {
  const rec = corpus.history_expn.get(pid.id as Documented<"history_expn">)
  return rec?.kind === "event-designator"
}

const docsShape: SchemaShape<"key" | "category"> = {
  type: "object",
  properties: {
    key: { type: "string" },
    category: categoryShape,
  },
  required: ["key"],
  additionalProperties: false,
}

export const docsToolDef: ToolDef = buildToolDef<"key" | "category">({
  name: "zsh_docs",
  prose: docsProse,
  shape: docsShape,
  outputSchema: mkOutputSchema({ mdBody: "required", feedback: "optional" }),
  execute: (corpus, input): DocsResult =>
    docs(corpus, input as unknown as DocsInput),
})

File: zshref-rs/src/tools/docs.rs
//! `zsh_docs` — zsh key → per-category resolved matches.
//!
//! Without `category`, walks `CLASSIFY_ORDER` and returns one match per
//! resolving category. Feedback (e.g. `NO_`-stripping → `input-negated`)
//! is forwarded from the per-category resolver.

use crate::corpus::{CLASSIFY_ORDER, Corpus, DocCategory};
use crate::resolver::{ResolvedHit, ResolverFeedback, resolve_in};
use crate::tools::envelope::Envelope;
use crate::tools::schema::{MatchShape, Shape, output_schema};
use crate::tools::{Field, Tool, ToolName, prose};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub fn tool(corpus: &Corpus) -> Tool {
    Tool::new(
        ToolName::Docs,
        prose::docs(corpus.index),
        vec![
            Field::required("key", prose::key(), Shape::Text),
            Field::optional(
                "category",
                prose::docs_category(corpus.index),
                Shape::Category,
            ),
        ],
        output_schema(
            &MatchShape {
                title: true,
                md_body: true,
                feedback: true,
                ..MatchShape::default()
            },
            corpus,
        ),
        run,
    )
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct Input {
    key: String,
    category: Option<DocCategory>,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
struct Match<'c> {
    category: DocCategory,
    id: &'c str,
    display: &'c str,
    title: &'c str,
    md_body: &'c str,
    #[serde(skip_serializing_if = "Option::is_none")]
    sub_kind: Option<&'c str>,
    #[serde(skip_serializing_if = "Option::is_none")]
    feedback: Option<ResolverFeedback>,
}

impl<'c> From<ResolvedHit<'c>> for Match<'c> {
    fn from(h: ResolvedHit<'c>) -> Self {
        Self {
            category: h.category,
            id: h.id,
            display: h.display,
            title: h.rec.title(),
            md_body: h.rec.md_body(),
            sub_kind: h.rec.sub_kind(),
            feedback: h.feedback,
        }
    }
}

/// In the category walk, `history_expn` counts only for event designators:
/// a bare word designator or modifier (`0`, `h`) is not a history token.
fn walk_admits(h: &ResolvedHit) -> bool {
    h.category.as_str() != "history_expn" || h.rec.sub_kind() == Some("event-designator")
}

fn run(input: Input, corpus: &Corpus) -> Result<Value> {
    let key = input.key.as_str();
    let matches: Vec<Match> = if key.trim().is_empty() {
        Vec::new()
    } else {
        match input.category {
            Some(cat) => resolve_in(corpus, cat, key)
                .map(Match::from)
                .into_iter()
                .collect(),
            None => CLASSIFY_ORDER
                .iter()
                .filter_map(|&cat| {
                    resolve_in(corpus, cat, key)
                        .filter(walk_admits)
                        .map(Match::from)
                })
                .collect(),
        }
    };
    let n = matches.len();
    Envelope::new(matches, n).to_value()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::corpus::load_corpus;
    use serde_json::json;

    #[test]
    fn walk_skips_history_modifiers_but_scoping_finds_them() {
        let corpus = load_corpus().expect("load_corpus");
        let docs = tool(&corpus);
        let categories = |input: Value| -> Vec<Value> {
            let out = docs.call(&input, &corpus).expect("docs");
            let matches = out["matches"].as_array().expect("matches");
            matches.iter().map(|m| m["category"].clone()).collect()
        };
        assert!(!categories(json!({"key": "h"})).contains(&json!("history_expn")));
        assert_eq!(
            categories(json!({"key": "h", "category": "history_expn"})),
            [json!("history_expn")]
        );
    }
}
